File: terminal_erp/crm/report/first_response_time_for_opportunity/first_response_time_for_opportunity.js
// Copyright (c) 2016, Terminal Framework Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

terminal_framework.query_reports["First Response Time for Opportunity"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			reqd: 1,
			default: terminal_framework.datetime.add_days(terminal_framework.datetime.nowdate(), -30),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			reqd: 1,
			default: terminal_framework.datetime.nowdate(),
		},
	],
	get_chart_data: function (_columns, result) {
		return {
			data: {
				labels: result.map((d) => d.creation_date),
				datasets: [
					{
						name: "First Response Time",
						values: result.map((d) => d.first_response_time),
					},
				],
			},
			type: "line",
			tooltipOptions: {
				formatTooltipY: (d) => {
					let duration_options = {
						hide_days: 0,
						hide_seconds: 0,
					};
					return terminal_framework.utils.get_formatted_duration(d, duration_options);
				},
			},
		};
	},
};

File: terminal_erp/crm/report/first_response_time_for_opportunity/first_response_time_for_opportunity.py
# Copyright (c) 2013, Terminal Framework Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt


import terminal_framework
from terminal_framework import _
from terminal_framework.query_builder.functions import Avg, Date
from pypika import Order


def execute(filters=None):
	columns = [
		{"fieldname": "creation_date", "label": _("Date"), "fieldtype": "Date", "width": 300},
		{
			"fieldname": "first_response_time",
			"fieldtype": "Duration",
			"label": "First Response Time",
			"width": 300,
		},
	]

	opportunity = terminal_framework.qb.DocType("Opportunity")
	creation_date = Date(opportunity.creation)
	data = (
		terminal_framework.qb.from_(opportunity)
		.select(
			creation_date.as_("creation_date"), Avg(opportunity.first_response_time).as_("avg_response_time")
		)
		.where(
			creation_date.between(filters.from_date, filters.to_date) & (opportunity.first_response_time > 0)
		)
		.groupby(creation_date)
		.orderby(creation_date, order=Order.desc)
		.run()
	)

	return columns, data
